### PR TITLE
[v10.1.x] Annotations: Split cleanup into separate queries and deletes to avoid deadlocks on MySQL

### PR DESCRIPTION
**Pull request by alexweav**
**What is this feature?**

Manual backport of https://github.com/grafana/grafana/pull/80329

### CODE REVIEW

File: pkg/services/annotations/annotationsimpl/xorm_store.go [hunk 1]
```diff
@@ -578,3 +653,8 @@ func (r *xormRepositoryImpl) executeUntilDoneOrCancelled(ctx context.Context, sq
 		}
 	}
 }
+
+type annotationTag struct {
```
Comment (alexweav) [inline pkg/services/annotations/annotationsimpl/xorm_store.go:657]: Backported from later versions; utility for reading tags with xorm, the tests depend on it.

File: pkg/services/annotations/annotationsimpl/xorm_store.go [hunk 2]
```diff
@@ -521,53 +522,127 @@ func (r *xormRepositoryImpl) validateTagsLength(item *annotations.Item) error {
 func (r *xormRepositoryImpl) CleanAnnotations(ctx context.Context, cfg setting.AnnotationCleanupSettings, annotationType string) (int64, error) {
 	var totalAffected int64
 	if cfg.MaxAge > 0 {
-		cutoffDate := time.Now().Add(-cfg.MaxAge).UnixNano() / int64(time.Millisecond)
-		deleteQuery := `DELETE FROM annotation WHERE id IN (SELECT id FROM (SELECT id FROM annotation WHERE %s AND created < %v ORDER BY id DESC %s) a)`
-		sql := fmt.Sprintf(deleteQuery, annotationType, cutoffDate, r.db.GetDialect().Limit(r.cfg.AnnotationCleanupJobBatchSize))
+		cutoffDate := timeNow().Add(-cfg.MaxAge).UnixNano() / int64(time.Millisecond)
```
Comment (alexweav) [inline pkg/services/annotations/annotationsimpl/xorm_store.go:525]: backport change: time.Now() -> timeNow()

File: pkg/services/annotations/annotationsimpl/cleanup_test.go [hunk 3]
```diff
@@ -223,22 +267,44 @@ func createTestAnnotations(t *testing.T, store db.DB, expectedCount int, oldAnno
...
+		return nil
+	})
+	require.NoError(t, err)
 }
 
 func settingsFn(maxAge time.Duration, maxCount int64) setting.AnnotationCleanupSettings {
 	return setting.AnnotationCleanupSettings{MaxAge: maxAge, MaxCount: maxCount}
 }
+
+func min(is ...int) int {
```
Comment (alexweav) [inline pkg/services/annotations/annotationsimpl/cleanup_test.go:299]: Wrote manually, 10.1 used an older version of Go where `min()` is not available.

**Review by JacobsonMT — APPROVED** [review]
LGTM

**Pull request merged**